---
title: "Plot_dimreducation_clustering"
author: "Jiawei Huang"
output: html_document
---

### Essential packages

```{r,warning = FALSE}
library(plyr)
library(reshape2)
library(ManiNetCluster)
library(ggplot2)
library(RColorBrewer)
library(cowplot)
library(stringr)
library(Seurat)
library(dplyr)
library(corrplot)
library(readxl)
library(cluster)

Dim_red = function(ef,gf,method,class,cellnames,d,k_NN,k_medoid){
  #e-feature
  X = apply(ef,2,scale)
  #g-feature
  Y=t(log10(gf+1))
  #Dim_red
  XY_corr=Correspondence(matrix=diag(nrow(X)))
  df=ManiNetCluster(X,Y,nameX='Ephys',nameY='Expr',corr=XY_corr,d=d,
                    method=method,k_NN=k_NN,k_medoids=k_medoid)
  df$ttype = rep(class,2)
  df$cellnames = rep(unlist(cellnames),2)
  return(df[,-1])
}
```

### Alignment using NMA,CCA and PCA

```{r}
ef = read.csv("data/efeature_filtered.csv",header = T,stringsAsFactors = F)
gf = read.csv("data/expMat_filtered.csv",header = T,stringsAsFactors = F)
match = read.csv('data/20200711_patchseq_metadata_mouse.csv',header=T,check.names = F,stringsAsFactors = F)
rownames(ef) = ef[,1];ef = ef[,-1]
rownames(gf) = gf[,1];gf = gf[,-1]
cellnames = rownames(ef)
match$genotype = sapply(match$t_type,function(x){na.omit(unlist(strsplit(x, " ")))[1]})
t_type = sapply(cellnames,FUN = function(x){match$genotype[match$transcriptomics_sample_id==x]})
n = nrow(ef)
method = c('linear manifold','cca','manifold warping','nonlinear manifold aln','nonlinear manifold warp')

### MA ###
NMA_res = Dim_red(ef,gf,method = method[4],class=t_type,cellnames =cellnames,d=3L,k_NN=2L,k_medoid=5L)
NMA_res_e = NMA_res[NMA_res$data=="Ephys",]
NMA_res_t= NMA_res[NMA_res$data=="Expr",]
### CCA ###
CCA_res = Dim_red(ef,gf,method = method[2],class=t_type,cellnames =cellnames,d=3L,k_NN=2L,k_medoid=5L)
CCA_res_e = CCA_res[CCA_res$data=="Ephys",]
CCA_res_t= CCA_res[CCA_res$data=="Expr",]
### PCA ###
#clustering only use gfs
PCA_res_t = prcomp(t(log10(gf+1)),scale=T,rank=3,retx=T)$x
PCA_res_t = data.frame(Val0 = as.numeric(PCA_res_t[,1]),Val1 = as.numeric(PCA_res_t[,2]),Val2 = as.numeric(PCA_res_t[,3]))
#clustering only use efs
PCA_res_e = prcomp(ef,scale=T,rank=3,retx=T)$x
PCA_res_e = data.frame(Val0 = as.numeric(PCA_res_e[,1]),Val1 = as.numeric(PCA_res_e[,2]),Val2 = as.numeric(PCA_res_e[,3]))
PCA_res = rbind(PCA_res_e[,1:3],PCA_res_t[,1:3])
head(NMA_res)

#Fig 1C
library(plot3D)
#NMA
NMA_plot = -rbind(apply(NMA_res[1:n,2:4],2,scale),apply(NMA_res[(n+1):(2*n),2:4],2,scale))
NMA_plot[1:n,1] = NMA_plot[1:n,1] + 0.5
points3D(x=NMA_plot[,1], y=NMA_plot[,2], z=NMA_plot[,3],pch = 19,cex=0.5,bty="g",ticktype = "detailed", theta = 40, phi = 10,
         xlab = "",ylab = "",zlab = "",box=T, axes=F,
         colvar = as.numeric(mapvalues(NMA_res$data,names(table(NMA_res$data)),1:2)),col = alpha(brewer.pal(6,"Spectral")[c(1,6)],c(0.8,0.8)),
         colkey = F)
#CCA
CCA_plot = rbind(apply(CCA_res[1:n,2:4],2,scale),apply(CCA_res[(n+1):(2*n),2:4],2,scale))
CCA_plot[1:n,1] = CCA_plot[1:n,1] + 0.5
points3D(x=CCA_plot[,1], y=CCA_plot[,2], z=CCA_plot[,3],pch = 19,cex=0.5,bty="g",ticktype = "detailed", theta = 40, phi = 10,
         xlab = "",ylab = "",zlab = "",box=T, axes=F,
         colvar = as.numeric(mapvalues(CCA_res$data,names(table(CCA_res$data)),1:2)),col = alpha(brewer.pal(6,"Spectral")[c(1,6)],c(0.8,0.8)),
         colkey = F)
#PCA
PCA_plot = rbind(apply(PCA_res_e[,1:3],2,scale),apply(PCA_res_t[,1:3],2,scale))
PCA_plot[1:n,1] = PCA_plot[1:n,1] + 0.5
points3D(x=PCA_plot[,1], y=PCA_plot[,2], z=PCA_plot[,3],pch = 19,cex=0.5,bty="g",ticktype = "detailed", theta = 40, phi = 10,
         xlab = "",ylab = "",zlab = "",box=T, axes=F,
         colvar = as.numeric(mapvalues(NMA_res$data,names(table(NMA_res$data)),1:2)),col = alpha(brewer.pal(6,"Spectral")[c(1,6)],c(0.8,0.8)),
         colkey = F)
```

### Clustering

```{r}
library(ClusterR)
#gmm optimal cluster (Fig S4)
opt_gmm = Optimal_Clusters_GMM(cbind(NMA_res[1:n,2:4],NMA_res[(n+1):(2*n),2:4]), max_clusters = 10,
                               criterion = "BIC",dist_mode = "eucl_dist", seed_mode = "random_subset",
                               km_iter = 50, em_iter = 100, var_floor = 1e-10,plot_data = T)
#NMA cluster
gmm = GMM(cbind(NMA_res[1:n,2:4],NMA_res[(n+1):(2*n),2:4]), 5, dist = "eucl_dist", "random_subset", 10, 50,seed=1)
gmm_cluster = predict_GMM(cbind(NMA_res[1:n,2:4],NMA_res[(n+1):(2*n),2:4]),
                          gmm$centroids, gmm$covariance_matrices, gmm$weights)$cluster_labels +1
#CCA cluster
gmm_cca = GMM(cbind(CCA_res[1:n,2:4],CCA_res[(n+1):(2*n),2:4]), 5, dist = "eucl_dist", "random_subset", 10, 50,seed=1)
gmm_cluster_cca = predict_GMM(cbind(CCA_res[1:n,2:4],CCA_res[(n+1):(2*n),2:4]),
                          gmm_cca$centroids, gmm_cca$covariance_matrices, gmm_cca$weights)$cluster_labels +1
#PCA cluster
gmm_pca = GMM(cbind(PCA_res_e,PCA_res_t), 5, dist = "eucl_dist", "random_subset", 10, 50,seed=1)
gmm_cluster_pca = predict_GMM(cbind(PCA_res_e,PCA_res_t),
                              gmm_pca$centroids, gmm_pca$covariance_matrices,gmm_pca$weights)$cluster_labels +1

#t-type & cluster (Fig S5)
group_ref = matrix(0,ncol = 5,nrow = 6)
rownames(group_ref) = names(table(NMA_res$ttype))
colnames(group_ref) = factor(paste0("Cluster",1:5),levels = paste0("Cluster",1:5))
for (i in 1:n){
  group_ref[NMA_res_e[i,]$ttype,gmm_cluster[i]] = group_ref[NMA_res_e[i,]$ttype,gmm_cluster[i]] + 1
}
corrplot(group_ref,tl.col = "black",tl.srt = 45,is.corr = FALSE,cl.lim=c(0,1500))

#t-type distribution (Fig 2A)
#e-features NMA
points3D(x=-NMA_res_e$Val0, y=-NMA_res_e$Val1, z=-NMA_res_e$Val2,pch = 19,cex=0.5,bty="g",ticktype = "detailed", theta = 40, phi = 10,
         xlab = "",ylab = "",zlab = "",
         colvar = as.numeric(mapvalues(t_type,names(table(t_type)),1:6)),col =alpha(brewer.pal(6,"Set1"),0.8),
         colkey = F)
#e-features PCA
points3D(x=PCA_res_e$Val0, y=PCA_res_e$Val1, z=PCA_res_e$Val2,pch = 19,cex=0.5,bty="g",ticktype = "detailed", theta = 40, phi = 10,
         xlab = "",ylab = "",zlab = "",
         colvar = as.numeric(mapvalues(t_type,names(table(t_type)),1:6)),col =alpha(brewer.pal(6,"Set1"),0.8),
         colkey = F)
#e-features CCA
points3D(x=CCA_res_e$Val0, y=CCA_res_e$Val1, z=CCA_res_e$Val2,pch = 19,cex=0.5,bty="g",ticktype = "detailed", theta = 40, phi = 10,
         xlab = "",ylab = "",zlab = "",
         colvar = as.numeric(mapvalues(t_type,names(table(t_type)),1:6)),col =alpha(brewer.pal(6,"Set1"),0.8),
         colkey = F)
colkey(at = 1:6,clim = c(0.5,6.5),alpha(brewer.pal(6,"Set1"),0.8),clab = "t-types", 
       addlines = F, length = 0.5, width = 0.5,
       labels = c("Lamp5","Pvalb","Serpinf1","Sncg","Sst","Vip"))
# #t-type distribution (Fig S2)
# #gene NMA
# points3D(x=-NMA_res_t$Val0, y=-NMA_res_t$Val1, z=-NMA_res_t$Val2,pch = 19,cex=0.5,bty="g",ticktype = "detailed", theta = 40, phi = 10,
#          xlab = "",ylab = "",zlab = "",
#          colvar = as.numeric(mapvalues(t_type,names(table(t_type)),1:6)),col =alpha(brewer.pal(6,"Set1"),0.8),
#          colkey = F)
# #gene PCA
# points3D(x=PCA_res_t$Val0, y=PCA_res_t$Val1, z=PCA_res_t$Val2,pch = 19,cex=0.5,bty="g",ticktype = "detailed", theta = 40, phi = 10,
#          xlab = "",ylab = "",zlab = "",
#          colvar = as.numeric(mapvalues(t_type,names(table(t_type)),1:6)),col =alpha(brewer.pal(6,"Set1"),0.8),
#          colkey = F)
# #gene CCA
# points3D(x=CCA_res_t$Val0, y=CCA_res_t$Val1, z=CCA_res_t$Val2,pch = 19,cex=0.5,bty="g",ticktype = "detailed", theta = 40, phi = 10,
#          xlab = "",ylab = "",zlab = "",
#          colvar = as.numeric(mapvalues(t_type,names(table(t_type)),1:6)),col =alpha(brewer.pal(6,"Set1"),0.8),
#          colkey = F)
```

### silhouette value

```{r}
#Fig 2B
silmat = data.frame(
  PCA = silhouette(as.numeric(mapvalues(t_type,names(table(t_type)),1:6)), dist = dist(PCA_res_e[,1:3]))[,3],
  CCA = silhouette(as.numeric(mapvalues(t_type,names(table(t_type)),1:6)), dist = dist(CCA_res_e[,2:4]))[,3],
  NMA = silhouette(as.numeric(mapvalues(t_type,names(table(t_type)),1:6)), dist = dist(NMA_res_e[,2:4]))[,3]
)

silmat = melt(silmat,variable.name = "Method", value.name = "Silhouette Value")
boxplot(`Silhouette Value`~Method,data=silmat,
        ylab="Silhouette Value", xlab="Cluster",outline =F,col ="white",ylim=c(-1,1))
tapply(silmat$`Silhouette Value`,INDEX = silmat$Method,mean)
```

### differential expressed genes & representive e features

```{r}
#different expressed genes 
diffg <- CreateSeuratObject(counts =log10(gf+1), min.cells = 3, min.features = 200)
all.genes <- rownames(diffg)
diffg <- ScaleData(diffg, features = all.genes)
Idents(diffg) = factor(paste0("Cluster ",gmm_cluster))
diffg.markers <- FindAllMarkers(diffg, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
top10 <- diffg.markers  %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
DoHeatmap(diffg, features = top10$gene,group.colors = brewer.pal(6,"Dark2")[1:5],angle = 30,size = 5) + NoLegend()
#write.csv(diffg.markers,"diff_genes.csv")
# for (i in 1:5){
#   cat("Cluster",i,"\n",diffg.markers$gene[diffg.markers$cluster==paste0("Cluster ",i)],"\n")
# }

#representive e-features
diffe <- CreateSeuratObject(counts =abs(t(ef)),min.cells = 1, min.features = 1)
all.genes <- rownames(diffe)
diffe <- ScaleData(diffe, features = all.genes)
Idents(diffe) = factor(gmm_cluster,levels = c(1:5))
diffe.markers <- FindAllMarkers(diffe, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
#diffe.markers
#write.csv(diffe.markers,"diff_ef.csv")

# #Fig 2D, S7
# diffe.markers$`-log10p` = -log10(diffe.markers$p_val_adj)
# p<-ggplot(data=diffe.markers[diffe.markers$cluster ==1,], aes(x = reorder(factor(stringr::str_wrap(gene,15)),`-log10p`), y=`-log10p`)) + labs(y="enrichment")+
#   geom_bar(stat="identity",fill = brewer.pal(6,"Spectral")[5],width=0.4)+ theme_minimal() + 
#   labs(y="-log10(adjustded p value)",x="") +
#   coord_flip() + 
#   theme(axis.text = element_text(size=12.5,lineheight=.7),
#         axis.title = element_text(size=11,face = "bold")) 
# p
```

